### ISSUE TITLE
design: capture FlushGraphParams into gen_graph — eliminate decoder→sampling GPU idle gap

### ISSUE THREAD

**Issue by kzjeef** [kzjeef/dash-infer#143]
# Design: Capture FlushGraphParams into Gen Graph

## Problem

nsys profiling of the double-buffered CUDA graph decode path revealed a **~2.3ms GPU idle gap** per decode step between the decoder graph and gen_graph (sampling):

```
GPU timeline (per decode step):

[== decoder graph 4.4ms ==]  ~~~2.3ms idle~~~  [gen_graph 15us]  ~~2.3ms idle~~
                                    ↑                                  ↑
                         FlushGraphParams GPU work            turnaround to next step
                         + cudaGraphLaunch API overhead
```

`FlushGraphParams` runs 3 GPU operations **eagerly** (outside the captured gen_graph), right before `cudaGraphLaunch`:

| # | Operation | What it does |
|---|-----------|-------------|
| 1 | `flush_batch_gencfg` | `cudaMemcpyAsync` H2D: packed gencfg buffer (9 fields × batch × 4B) |
| 2 | `fill_max_dec_ids_gpu` | Loop of `cudaMemcpyAsync` D2D per request + `cudaMemsetAsync` for padding rows (repetition penalty input) |
| 3 | `sample_states_` H2D | `cudaMemcpyAsync` H2D: curandState device pointers array |

The GPU work itself is small (~15us), but the CPU-side overhead of `UpdateGraphParams` + `FlushGraphParams` function calls + `cudaGraphLaunch` API dominates the gap, keeping the GPU idle for ~2.3ms.

## Solution: capture flush ops into the gen_graph

The same pattern already works for the **decoder graph**: RotaryOp, SpanAttnOp, and DecOptEmbeddingOp all have H2D memcpy nodes captured inside graph segments. `UpdateGraphParams` writes fresh data to stable pinned host addresses before replay, and the captured memcpy nodes read from those addresses.

For gen_graph, we do the same:

```
Before (eager flush + graph replay):
  CPU: UpdateGraphParams() → FlushGraphParams() → cudaGraphLaunch(gen_graph)
  GPU: [idle waiting for CPU] [flush H2D+D2D] [gen_graph kernels]

After (flush captured inside graph):
  CPU: UpdateGraphParams() → cudaGraphLaunch(gen_graph)
  GPU: [idle ~33us] [gen_graph: flush H2D → kernel → sample_states H2D → sampling kernels]
```

### Sub-problem: fill_max_dec_ids uses variable D2D source pointers

`fill_max_dec_ids_gpu` does a loop of `cudaMemcpyAsync(D2D)` where each iteration's source address comes from `gen_ctx->request->interim.at("generated_ids_gpu")`. The source pointers change per step (as tokens are generated). **CUDA graph memcpy nodes have fixed addresses** — they can't replay with different source pointers.

**Fix — indirection table pattern** (same as SpanAttnOp uses for span pointer arrays):

```
                    ┌─── UpdateGraphParams (host-only, overlaps with GPU) ───┐
                    │                                                         │
                    │  fill_src_ptrs_host_[0] = generated_ids_gpu[0].ptr     │
                    │  fill_src_ptrs_host_[1] = generated_ids_gpu[1].ptr     │
                    │  fill_sizes_host_[0]    = generated_ids_gpu[0].bytes   │
                    │  fill_sizes_host_[1]    = generated_ids_gpu[1].bytes   │
                    │  (padding slots get nullptr / 0)                        │
                    └─────────────────────────────────────────────────────────┘
                                            │
                          ┌─── gen_graph (captured, GPU) ───────────────────┐
                          │                                                  │
                          │  ① H2D: fill_src_ptrs_host_ → device            │
                          │  ② H2D: fill_sizes_host_    → device            │
                          │  ③ FillMaxDecIdsBatched kernel:                  │
                          │     for each block b:                            │
                          │       if src_ptrs[b] != nullptr:                 │
                          │         copy src_ptrs[b] → max_dec_ids[b*L]     │
                          │       else:                                      │
                          │         zero max_dec_ids[b*L .. b*L+L]          │
                          │  ④ H2D: packed_gencfg_host_ → device            │
                          │  ⑤ H2D: sample_states_host_ → device            │
                          │  ⑥ process_logits kernel                         │
                          │  ⑦ TopK / TopP / Sample kernels                  │
                          └──────────────────────────────────────────────────┘
```

The kernel uses `src_ptrs[b] != nullptr` to distinguish real vs padding slots — no baked-in `actual_batch` count, so it's safe for graph replay when the batch composition changes between steps.

### Sub-problem: pageable host memory causes ~100x graph replay slowdown

Initial implementation used pageable `std::vector` for `packed_gencfg_host_` and `sample_states_host_`. When captured as H2D memcpy graph nodes, replay was catastrophically slow: gen_graph duration went from 15us to **4300us**.

**Root cause**: CUDA graph replay of H2D memcpy nodes from pageable memory requires the driver to allocate internal staging buffers and perform synchronous copies, adding massive per-node overhead.

**Fix**: Convert all host buffers used as graph-captured H2D sources to **pinned memory** (`cudaMallocHost` via `AsTensorFlags::pinned_host_mem`):

| Buffer | Before | After |
|--------|--------|-------|
| `packed_gencfg_host_` | `std::vector<char>` | pinned `AsTensor` |
| `sample_states_host_` | `std::vector<void*>` | pinned `AsTensor` |
| `fill_src_ptrs_host_` | (new) | pinned `AsTensor` |
| `fill_sizes_host_` | (new) | pinned `AsTensor` |

After this fix, gen_graph duration: 15us → **46us** (+31us for flush ops). Correct.

## Address stability verification

All addresses in captured graph nodes must be stable within a graph plan's lifetime:

```
Pinned host (H2D source — written by UpdateGraphParams, read by graph replay):
  packed_gencfg_host_->GetDataPtr()    alloc in Init(), fixed    ✓
  sample_states_host_->GetDataPtr()    alloc in Init(), fixed    ✓
  fill_src_ptrs_host_->GetDataPtr()    alloc in Init(), fixed    ✓
  fill_sizes_host_->GetDataPtr()       alloc in Init(), fixed    ✓

Device (H2D dest / kernel operands — captured addresses):
  packed_gencfg_device_->GetDataPtr()  alloc in first Reshape    ✓ (fixed within bucket)
  sample_states_->GetDataPtr()         alloc in Init(), fixed    ✓
  fill_src_ptrs_device_->GetDataPtr()  alloc in Init(), fixed    ✓
  fill_sizes_device_->GetDataPtr()     alloc in Init(), fixed    ✓
  max_dec_ids_->GetDataPtr()           alloc in Init(), fixed    ✓

Device (indirection table values — change per step, read by kernel):
  generated_ids_gpu->GetDataPtr()      alloc at Shape{1,max_len} ✓ (stable base ptr)
```

### New kernel: `FillMaxDecIdsBatched` (`copy_var_array.cu`)

```
Grid:  padded_batch blocks (one per batch row)
Block: 256 threads

Per block b:
  if src_ptrs[b] != nullptr:
    count = min(sizes[b] / 8, max_length)   // clamp to row size
    copy count int64s from src_ptrs[b] to max_dec_ids[b * max_length]
  else:
    zero max_dec_ids[b * max_length .. +max_length]
```

### New method: `RunFlushAndSampleGPU` (`generate_op.cpp`)

Sequences all flush + sampling operations into one capturable stream:

```cpp
AsStatus RunFlushAndSampleGPU(runtime_ctx):
  // These become captured graph nodes:
  flush_batch_gencfg()                    // gencfg H2D (pinned → device)
  cudaMemcpyAsync(src_ptrs H2D)           // indirection table H2D
  cudaMemcpyAsync(sizes H2D)              // indirection table H2D
  FillMaxDecIdsBatchedLauncher()           // batched D2D via table
  cudaMemcpyAsync(sample_states H2D)      // sample states H2D
  RunSampleGPU()                          // process_logits + TopK + Sample
```

### Modified: `UpdateGraphParams` (`generate_op.cpp`)

Added two new host-only sections (safe to overlap with GPU):

```cpp
// Fill indirection table for graph-captured fill_max_dec_ids
for i in [0, capped_batch):
  if i < actual_batch:
    src_ptrs[i] = gen_ctx[i]->generated_ids_gpu->GetDataPtr()
    sizes[i]    = gen_ctx[i]->generated_ids_gpu->GetSizeInByte()
  else:
    src_ptrs[i] = nullptr   // kernel will zero this row
    sizes[i]    = 0
```

### Modified: `model.cpp` gen_graph capture + replay

```
Capture (first time):
  cudaStreamBeginCapture(stream)
  gen_op->RunFlushAndSampleGPU(ctx)    // was: RunSampleGPU
  cudaStreamEndCapture(stream, &graph)

Replay (subsequent steps):
  gen_op->UpdateGraphParams(ctx)       // host-only: fill pinned buffers
  // gen_op->FlushGraphParams(ctx)     // REMOVED — now inside graph
  cudaGraphLaunch(gen_graph_exec, stream)
```

## Results (Qwen2.5-7B-Instruct, batch=1, H200 GPU 7, nsys)

```
                          Baseline      Optimized       Delta
Decoder graph avg         9069 us       4431 us         -51%
Gen graph avg               15 us         46 us         +31 us
Decoder→Gen gap           2363 us         33 us         -98.6%
Gen→Decoder gap           2353 us         89 us         -96.2%
Throughput                73.8 tok/s    213 tok/s       +189%
```

The decoder graph halving (9069→4431us) is because the double-buffer CPU/GPU overlap now actually works — without the 2.3ms flush gap stalling the pipeline, the CPU can prepare the next step's graph while the current step's decoder is still executing.

## Implementation: PR #140
